### PR TITLE
feat: [IC-272] add dfx canister logs support

### PR DESCRIPTION
**Pull request by maksymar**
# Description

This PR adds a new canister sub-command `logs` to fetch the canister's logs, eg `dfx canister logs <canister_id>`. When formatting log entries it tries to guess if the content can be converted into utf8 string or not and then is printed as an array of bytes.

# How Has This Been Tested?

Regular CI and a formatting test.

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: src/dfx/src/commands/canister/tail.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    let canister_id_store = env.get_canister_id_store()?;
+
+    let canister_id = Principal::from_text(callee_canister)
+        .or_else(|_| canister_id_store.get(callee_canister))?;
+
+    fetch_root_key_if_needed(env).await?;
+
+    let logs = canister::get_canister_logs(env, canister_id, call_sender).await?;
+
+    info!(log, "Canister tail call result for {canister_id}: {logs}");
```
[the remark this answers is not in the document]
Comment (maksymar) [reply src/dfx/src/commands/canister/tail.rs]: thank you! switched to only displaying the logs.

File: src/dfx/src/commands/canister/tail.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    assert_eq!(
+        format_canister_logs(logs),
+        vec![
+            "[42. 2021-05-06 19:17:10.000]: Some text message".to_string(),
+            "[43. 2021-05-06 19:17:10.001]: [1, 2, 3]".to_string(),
+        ],
+    );
+}
+
+pub async fn exec(env: &dyn Environment, opts: TailOpts, call_sender: &CallSender) -> DfxResult {
```
Comment (ericswanson-dfinity) [inline src/dfx/src/commands/canister/tail.rs:72]: I'd suggest calling this `dfx canister logs`, even if its default semantics are "tail" (meaning to return the last N log records).  What are the semantics?

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
Please add a section in https://github.com/dfinity/sdk/blob/master/docs/cli-reference/dfx-canister.md

File: src/dfx/src/commands/canister/tail.rs (hunk 2 above)
Comment (maksymar) [reply src/dfx/src/commands/canister/tail.rs:72]: @ielashi PTAL

Comment (maksymar) [reply src/dfx/src/commands/canister/tail.rs:72]: For reference here's the design doc: https://docs.google.com/document/d/10zx2qybpK9lmYBUn4r-AbNkcfA_k_zpM7XQ0yfyzBRc

File: src/dfx/src/commands/canister/tail.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                timestamp_nanos: 1620328630001000000,
+                content: vec![1, 2, 3],
+            },
+        ],
+    };
+    assert_eq!(
+        format_canister_logs(logs),
+        vec![
+            "[42. 2021-05-06 19:17:10.000]: Some text message".to_string(),
+            "[43. 2021-05-06 19:17:10.001]: [1, 2, 3]".to_string(),
```
Comment (ielashi) [inline src/dfx/src/commands/canister/tail.rs:67]: I agree with @adamspofford-dfinity's comment thow showing these in a hex string would be easier to work with than decimals. Perhaps also adding a prefix to indicate that these are raw bytes would help as well.

```suggestion
            "[43. 2021-05-06 19:17:10.001] (bytes): 0x010203".to_string(),
```

File: src/dfx/src/commands/canister/tail.rs (hunk 2 above)
Comment (ielashi) [reply src/dfx/src/commands/canister/tail.rs:72]: That makes sense to me. I am sure we'll want to add more functionality beyond just `tail` in the future, so having them all be part of the `dfx canister logs` command SGTM.

Comment (maksymar) [reply src/dfx/src/commands/canister/tail.rs]: Done.

File: src/dfx/src/commands/canister/tail.rs (hunk 3 above)
Comment (maksymar) [reply src/dfx/src/commands/canister/tail.rs]: Done.

File: src/dfx/src/commands/canister/tail.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                .expect("invalid timestamp");
+
+            let message = if let Ok(s) = String::from_utf8(r.content.clone()) {
+                if format!("{s:?}").contains("\\u{") {
+                    format!("{:?}", r.content)
+                } else {
+                    s
+                }
+            } else {
+                format!("{:?}", r.content)
```
[the remark this answers is not in the document]
Comment (maksymar) [reply src/dfx/src/commands/canister/tail.rs]: Done.

File: src/dfx/Cargo.toml [hunk 5]
```diff
@@ -40,6 +40,7 @@ byte-unit = { workspace = true, features = ["serde"] }
 bytes.workspace = true
 candid = { workspace = true }
 candid_parser = { workspace = true, features = ["random", "assist"] }
+chrono = "0.4.34"
```
[the remark this answers is not in the document]
Comment (maksymar) [reply src/dfx/Cargo.toml]: thank you! done.

File: docs/cli-reference/dfx-canister.md [hunk 6]
```diff
@@ -530,6 +531,31 @@ For example, you can specify a testnet URL by running a command similar to the f
...
+
+``` bash
+dfx canister logs hello_world
+```
+
+The command displays output similar to the following:
+
+``` log
+[42. 2021-05-06T19:17:10.000000001Z]: Some text message
+[43. 2021-05-06T19:17:10.000000002Z]: (bytes) 0xc0ffee
```
Comment (ielashi) [inline docs/cli-reference/dfx-canister.md:556]: Output LGTM

**Review by ielashi — APPROVED** [review]

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**